app/models/workers_scaling/experiment_resources_interface.rb

```diff
@@ -0,0 +1,221 @@
+          .map(&:sm_uuid)
+          .concat(starting_workers + initializing_workers)
+      rescue InvalidCredentialsError, NoCredentialsError
+        # TODO inform user about credentials error
+        raise
+      end
+    end
+
+    ##
+    # Marks worker with given sm_uuid to be deleted after finishing given number of simulations
+    # If overwrite is set to false, worker will not be marked if simulations_left field is already set,
+    # otherwise new number will be set in all cases
+    # Using overwrite=true may result in unintentional resetting simulations_left field,
+    # effectively prolonging lifetime of worker, therefore default value is false
+    def limit_worker_simulations(sm_uuid, simulations_left, overwrite=false)
+      worker = get_worker_record_by_sm_uuid(sm_uuid)
+      unless worker.nil?
+        worker.simulations_left = simulations_left if worker.simulations_left.blank? or overwrite
+        worker.save
+      end
+    end
+
+    ##
+    # Stops worker after completing its current simulation execution
+    def soft_stop_worker(sm_uuid)
+      limit_worker_simulations(sm_uuid, 1)
+    end
+
+    ##
+    # Returns list of workers records for infrastructure
+    def get_workers_records_list(infrastructure, params = {})
+      get_workers_records_cursor(infrastructure, params).to_a
+    end
+
+    ##
+    # Returns workers records count for infrastructure
+    def get_workers_records_count(infrastructure, params = {})
+      get_workers_records_cursor(infrastructure, params).count
+    end
+
+    ##
+    # Returns overall Workers count for Experiment matching given params
+    def count_all_workers(params = {})
+      get_available_infrastructures
+          .flat_map { |infrastructure| get_workers_records_count(infrastructure, params) }
+          .reduce(0) { |sum, count| sum + count }
+    end
+
+    ##
+    # Returns worker record for given sm_uuid
+    def get_worker_record_by_sm_uuid(sm_uuid)
+      get_available_infrastructures.flat_map do |infrastructure|
+        get_workers_records_cursor(infrastructure, cond: {sm_uuid: sm_uuid}).first
+      end .first
+    end
+
+    ##
+    # Yields workers for given records
+    # Usage:
+    #   yield_workers(records) do |worker|
+    #      worker.<some method>
+    #   end
+    # Workers commands: restart, stop, destroy_record
+    def yield_workers(records)
+      records.each do |record|
+        get_facade_for(record.infrastructure).yield_simulation_manager(record) {|worker| yield worker}
+      end
+    end
+
+    private
+
+    ##
+    # Checks whether all fields from narrower are equal with corresponding fields from wider
+    # when exact flag is set to false. Performs full comparison when exact flag is true.
+    # By default exact flag is set to false
+    # Returns true when infrastructures are equal, false otherwise.
+    def infrastructures_equal?(narrower, wider, exact=false)
```

Comment:
Tutaj znowu doczepię się używania nazwy "infrastructure". Chodzi o Hash (lub jak mieliście refaktorować - klasę "InfrastructureId") z konfiguracją zasobu na infrastrkturze? Lepiej już tutaj by pasowało ``infrastructure_configs_equal?``. Nie chciałbym mieszania pojęć w systemie. Infrastruktura to byt, który udostępnia zasoby w różnych konfiguracjach. Tutaj mamy do czynienia z tymi konfiguracjami.